test/model/src/driver/kv_workload_generator.cpp

```diff
@@ -234,86 +234,6 @@ kv_workload_generator::sequence_traversal::complete_one(sequence_state *s)
-            err << "The stable timestamp must not be smaller than the oldest timestamp: " << t
-                << " < " << oldest << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-        stable = t;
-    }
-
-    if (std::holds_alternative<operation::set_oldest_timestamp>(op)) {
-        timestamp_t t = std::get<operation::set_oldest_timestamp>(op).oldest_timestamp;
-        if (t < oldest) {
-            std::ostringstream err;
-            err << "The oldest timestamp went backwards: " << oldest << " -> " << t << " (sequence "
-                << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-        if (t > stable && stable != k_timestamp_none) {
-            std::ostringstream err;
-            err << "The oldest timestamp must not be later than the stable timestamp: " << t
-                << " > " << stable << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-        oldest = t;
-    }
-
-    if (std::holds_alternative<operation::prepare_transaction>(op)) {
-        timestamp_t t = std::get<operation::prepare_transaction>(op).prepare_timestamp;
-        if (t < stable) {
-            std::ostringstream err;
-            err << "Prepare timestamp is before the stable timestamp: " << t << " < " << stable
-                << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-    }
-
-    if (std::holds_alternative<operation::set_commit_timestamp>(op)) {
-        timestamp_t t = std::get<operation::set_commit_timestamp>(op).commit_timestamp;
-        if (t < stable) {
-            std::ostringstream err;
-            err << "Commit timestamp is before the stable timestamp: " << t << " < " << stable
-                << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-    }
-
-    if (std::holds_alternative<operation::commit_transaction>(op)) {
-        timestamp_t t = std::get<operation::commit_transaction>(op).commit_timestamp;
-        if (t < stable) {
-            std::ostringstream err;
-            err << "Commit timestamp is before the stable timestamp: " << t << " < " << stable
-                << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-        t = std::get<operation::commit_transaction>(op).durable_timestamp;
-        if (t < stable && t != k_timestamp_none) {
-            std::ostringstream err;
-            err << "Durable timestamp is before the stable timestamp: " << t << " < " << stable
-                << " (sequence " << sequence.seq_no() << ")" << std::endl;
-            throw model_exception(err.str());
-        }
-    }
-}
-
 /*
  * kv_workload_generator::assign_timestamps --
  *     Assign timestamps to operations in a sequence.
```

Comment:
A possible efficiency improvement.

Is the std::holds_alternative<operation::commit_transaction>(op) expensive? If so new lines 255 - 262 could be improved by two ideas:

1. Arrange the order of the std::holds_alternative<operation::whatever>(op) from most frequent to least frequent operation.
2. Rearrange the code so the second check of prepare_transaction is optimized away.

if (std::holds_alternative<operation::commit_transaction>(op) ||
     std::holds_alternative<operation::prepare_transaction>(op) ||
     std::holds_alternative<operation::set_commit_transaction>(op) ||
     std::holds_alternative<operation::set_oldest_timestamp>(op) ||
     std::holds_alternative<operation::set_stable_timestamp>(op)) {
         if std::holds_alternative<operation::prepare_transaction>(op))
             prepared = true;
         timestamped_ops.push_back(&op);
    }

I think an optimizing compiler after checking std::holds_alternative<operation::prepare_transaction>(op) will jump directly to prepared = true skipping the same check again.

Also the compiler for the other operations might jump to timestamped_ops.push_back(&op); skipping the if std::holds_alternative<operation::prepare_transaction>(op)). It would do this if it knew that only one of these checks could be true.